### PR TITLE
fix: disable bash completion if double dash is included in arguments (v2)

### LINKED ISSUE #1932: Shell completion works wrongly even if double dash `--` are included in arguments
## My urfave/cli version is

_**v2.27.2**_

## Checklist

- [x] Are you running the latest v2 release? The list of releases is [here](https://github.com/urfave/cli/releases).
- [ ] Did you check the manual for your release? The v2 manual is [here](https://github.com/urfave/cli/blob/main/docs/v2/manual.md)
- [x] Did you perform a search about this problem? Here's the [GitHub guide](https://help.github.com/en/github/managing-your-work-on-github/using-search-to-filter-issues-and-pull-requests) about searching.

## Dependency Management

- My project is using go modules.

## Describe the bug

I created a GitHub repository to reproduce the issue.
Please read README.md.

https://github.com/suzuki-shunsuke/reproduce-urfave-cli-shell-completion-issue

## To reproduce

Describe the steps or code required to reproduce the behavior

https://github.com/suzuki-shunsuke/reproduce-urfave-cli-shell-completion-issue

## Observed behavior

What did you see happen immediately after the reproduction steps
above?

https://github.com/suzuki-shunsuke/reproduce-urfave-cli-shell-completion-issue

## Expected behavior

What would you have expected to happen immediately after the
reproduction steps above?

https://github.com/suzuki-shunsuke/reproduce-urfave-cli-shell-completion-issue

## Additional context

If the issue relates to a specific open source GitHub repo, please
link that repo here.

If you can reproduce this issue with a public CI system, please
link a failing build here.

- Original issue: https://github.com/lintnet/lintnet/issues/528

This issue is related to my OSS project [aqua](https://aquaproj.github.io/), which is a CLI version manager written in Go.
aqua is using urfave/cli/v2 and is enabling shell completion.
aqua is a CLI version manager, and when a tool managed by aqua is executed, aqua intercepted the execution and executes the tool via `aqua exec -- <command> [<argument> ...]`.
For example, if a tool `lintnet` is managed by aqua and `lintnet --help` is executed, actually `aqua exec -- lintnet --help` is executed.

For detail, please see the document. https://aquaproj.github.io/docs/reference/lazy-install/#how-does-lazy-install-work

This issue causes a problem that shell completion doesn't work well if a tool managed by aqua is also built with urfave/cli/v2.

## Want to fix this yourself?

We'd love to have more contributors on this project! If the fix for
this bug is easily explained and very small, feel free to create a
pull request for it.

Yes.

## Run `go version` and paste its output here

```
# paste `go version` output in here
go version go1.22.4 darwin/arm64
```

## Run `go env` and paste its output here

```
# paste `go env` output in here
GO111MODULE=''
GOARCH='arm64'
GOBIN=''
GOCACHE='/Users/shunsukesuzuki/Library/Caches/go-build'
GOENV='/Users/shunsukesuzuki/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMODCACHE='/Users/shunsukesuzuki/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='darwin'
GOPATH='/Users/shunsukesuzuki/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/Users/shunsukesuzuki/.local/share/aquaproj-aqua/pkgs/http/golang.org/dl/go1.22.4.darwin-arm64.tar.gz/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/Users/shunsukesuzuki/.local/share/aquaproj-aqua/pkgs/http/golang.org/dl/go1.22.4.darwin-arm64.tar.gz/go/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.22.4'
GCCGO='gccgo'
AR='ar'
CC='clang'
CXX='clang++'
CGO_ENABLED='1'
GOMOD='/Users/shunsukesuzuki/repos/src/github.com/suzuki-shunsuke/reproduce-urfave-cli-shell-completion-issue/go.mod'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/fc/1bgyy3_d3x90m_t04qbw5f8m0000gn/T/go-build2726066441=/tmp/go-build -gno-record-gcc-switches -fno-common'
```

### PR DESCRIPTION
**Pull request by suzuki-shunsuke**
## What type of PR is this?

_(REQUIRED)_

- bug

## What this PR does / why we need it:

_(REQUIRED)_

This pull request disables shell completion if double dash `--` is included in arguments.
Arguments after double dash `--` are not flags but positional arguments.

https://unix.stackexchange.com/a/11382

## Which issue(s) this PR fixes:

_(REQUIRED)_

Fixes #1932

## Special notes for your reviewer:

_(fill-in or delete this section)_

## Testing

_(fill-in or delete this section)_

## Release Notes

_(REQUIRED)_

```release-note

```

### CODE REVIEW

**Comment by suzuki-shunsuke** [comment]
This pull request is for urfave/cli v2.

- PR for v3: https://github.com/urfave/cli/pull/1933

File: help_test.go [hunk 1]
```diff
@@ -1714,3 +1712,65 @@ GLOBAL OPTIONS:
...
+			app: &App{
+				EnableBashCompletion: true,
+			},
+			wantShellCompletion: true,
+			wantArgs:            []string{"foo"},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
```
Comment (dearchap) [inline help_test.go:1762]: is this needed ?

**Review by dearchap — APPROVED** [review]

**Comment by dearchap** [comment]
@suzuki-shunsuke Thank you for the PR. Can you make this PR for v3 as well ?

File: help_test.go (hunk 1 above)
Comment (suzuki-shunsuke) [reply help_test.go:1762]: If we remove this, `go vet` fails.

```console
$ go vet ./...
# github.com/urfave/cli/v2
# [github.com/urfave/cli/v2]
./help_test.go:1765:52: loop variable tt captured by func literal
./help_test.go:1765:60: loop variable tt captured by func literal
./help_test.go:1766:7: loop variable tt captured by func literal
./help_test.go:1768:23: loop variable tt captured by func literal
./help_test.go:1770:26: loop variable tt captured by func literal
./help_test.go:1772:12: loop variable tt captured by func literal

$ echo $?
1
```

If we update go directive to 1.22 or later, we can remove this.

https://tip.golang.org/doc/go1.22

https://github.com/urfave/cli/blob/8e2384caa9350d1cd514808821c181f20afd48f0/go.mod#L3

But I don't think this change is the scope of this pull request, so I didn't do that.

**Comment by suzuki-shunsuke** [comment]
> Can you make this PR for v3 as well ?

- https://github.com/urfave/cli/pull/1933

**Comment by suzuki-shunsuke** [comment]
@dearchap
Do you guys have any plan to release v2.27.3?

**Comment by dearchap** [comment]
I'll do it by tomorrow

**Comment by suzuki-shunsuke** [comment]
Thank you so much!

**Comment by suzuki-shunsuke** [comment]
Thank you for releasing v2.27.3!
https://github.com/urfave/cli/releases/tag/v2.27.3

**Pull request merged**